### ISSUE TITLE
nix-env --install (no args) behavior undocumented (and unwanted?)

### ISSUE THREAD

**Issue by nomeata** [NixOS/nix#3574]
I explained `nix-env --install` to a colleage of mine, new to nix, and he just ran it – as such:
```
$ nix-env --install
warning: there are multiple derivations named 'FlameGraph-2019-02-16'; using the first one
warning: there are multiple derivations named 'M33-Linux-2016-06-23'; using the first one
warning: there are multiple derivations named 'NetworkManager-fortisslvpn-gnome-1.2.10'; usi
…
25903 lines later
…
installing 'zziplib-0.13.69'
installing 'zzuf-0.15'
error: Package ‘1password-0.7.0’ in /nix/store/mjybh6jf1inil9bh9921j9vcskhkabl9-nixpkgs-20.09pre213040.f77e057cda6/nixpkgs/pkgs/applications/misc/1password/default.nix:37 has an unfree license (‘unfree’), refusing to evaluate.

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.

(use '--show-trace' to show detailed location information)
```

That’s confusing! I checked `nix-env --help`, and from reading that I would have expected the `args` to be required, and `nix-env --install` ideally giving a helpful error message. 

But instead it seems to try to install all of nixpkgs – which is hardly useful.

So either
 * `nix-env --install` prints a helpful error message or
 * (at least) the manpage explains what happens.

I’d advocate for the former, of course.

(This is with nix 2.3.1 and 2.3.4)

**Comment by jakobrs** [comment]
Note that you're using this functionality when you run `nix-env -if default.nix` or `nix-env -if https://example.com/path/to/archive.tar.gz`. If `nix-env` is not given any "target", it installs everything, recursively. That is, if the file (`default.nix` or `https://example.com/path/to/archive.tar.gz` or in this case, `<nixpkgs>`) evaluates to a derivation, it installs that derivation. If it evaluates to a list or attribute set of derivations, it installs each one.

Example:
```
$ cat >a.nix <<EOF
with import <nixpkgs> {};

[
  hello
  {
    a = vim;
    b = tmux;
  }
]
EOF
$ nix-env -if a.nix --dry-run
(dry run; not doing anything)
installing 'hello-2.10'
installing 'tmux-3.1b'
installing 'vim-8.2.0510'
...
```

**Comment by jakobrs** [comment]
Note that you're using this functionality when you run `nix-env -if default.nix` or `nix-env -if https://example.com/path/to/archive.tar.gz`. If `nix-env` is not given any "target", it installs everything, recursively. That is, if the file (`default.nix` or `https://example.com/path/to/archive.tar.gz` or in this case, `<nixpkgs>`) evaluates to a derivation, it installs that derivation. If it evaluates to a list or attribute set of derivations, it installs each one.

Example:
```
$ cat >a.nix <<EOF
with import <nixpkgs> {};

[
  hello
  {
    a = vim;
    b = tmux;
  }
]
EOF
$ nix-env -if a.nix --dry-run
(dry run; not doing anything)
installing 'hello-2.10'
installing 'tmux-3.1b'
installing 'vim-8.2.0510'
...
```

**Comment by nomeata** [comment]
The manpage explicitly writes
```
       To install all derivations in the Nix expression foo.nix:

           $ nix-env -f ~/foo.nix -i '.*'
```
to at least the docs seem to advice against just omitting `args` when you want all derivations.

I understand that this corner of possible flag uses (no `args`, default `-f`) assignemnt makes sense conceptually. But it is still is pretty bad for everyone who just runs `nix-env i` for some reason. And given that there is the (documented) `.*` way, maybe simply erroring with a helpful error message when _no_ args and _no_ -f is given would be nicer?

**Comment by jakobrs** [comment]
I agree that maybe `nix-env -i` specifically (without `-f`) could be a special case. And if people really want to try it, they can just add `.*`.

fwiw the manual does at least at one point use `-i` without `args`, in the release notes for Nix 0.8, when introducing `nix-env -ir`.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
